File: app/src/main/java/io/github/anvell/stackoverview/view/MainActivity.java
package io.github.anvell.stackoverview.view;

import android.animation.LayoutTransition;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.provider.Settings;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.LinearLayout;

import com.google.android.gms.ads.MobileAds;

import io.github.anvell.stackoverview.R;
import io.github.anvell.stackoverview.databinding.ActivityMainBinding;
import io.github.anvell.stackoverview.enumeration.ActiveScreen;
import io.github.anvell.stackoverview.model.QuestionDetails;
import io.github.anvell.stackoverview.viewmodel.MainViewModel;

import static android.arch.lifecycle.Lifecycle.State.RESUMED;
import static io.github.anvell.stackoverview.enumeration.ActiveScreen.DETAILS;
import static io.github.anvell.stackoverview.enumeration.ActiveScreen.SEARCH;

public class MainActivity extends AppCompatActivity implements SearchView.OnQueryTextListener {

    private MainViewModel viewModel;
    private SearchView searchView;
    private Toolbar toolbar;
    private ActivityMainBinding binding;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        toolbar = (Toolbar)binding.toolbarView;
        setSupportActionBar(toolbar);

        ActionBar supportActionBar = getSupportActionBar();
        if(supportActionBar != null) {
            getSupportActionBar().setCustomView(searchView);
            getSupportActionBar().setDisplayShowCustomEnabled(true);
        }

        viewModel = ViewModelProviders.of(this).get(MainViewModel.class);

        initObservers();
        setupHomeIcon();

        if (savedInstanceState == null) {
            initFragments();
        }
    }

    @Override
    public boolean onSupportNavigateUp() {
        invalidateOptionsMenu();
        if (getSupportFragmentManager().getBackStackEntryCount() > 0) {
            getSupportFragmentManager().popBackStack();
        }
        setupHomeIcon();
        return super.onSupportNavigateUp();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        this.onNavigateUp();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);

        MenuItem favorite = menu.findItem(R.id.mi_favorite);
        favorite.setVisible(viewModel.activeScreen.getValue() == DETAILS);

        QuestionDetails question = viewModel.selectedQuestion.getValue();
        if(question != null) {
            favorite.setIcon(getResources()
                            .getDrawable(question.isFavorite?
                                         R.drawable.ic_star :
                                         R.drawable.ic_star_hollow));
            favorite.setChecked(question.isFavorite);
        }

        MenuItem searchItem = menu.findItem(R.id.mi_search_view);
        searchView = (SearchView) searchItem.getActionView();
        searchView.setMaxWidth(Integer.MAX_VALUE);
        LinearLayout searchBar = searchView.findViewById(R.id.search_bar);
        searchBar.setLayoutTransition(new LayoutTransition());

        searchView.setOnQueryTextListener(this);

        searchItem.setOnActionExpandListener(new MenuItem.OnActionExpandListener() {
            @Override
            public boolean onMenuItemActionExpand(MenuItem item) {
                return true;
            }

            @Override
            public boolean onMenuItemActionCollapse(MenuItem item) {
                if (viewModel.activeScreen.getValue() == SEARCH) {
                    if (getSupportFragmentManager().getBackStackEntryCount() > 0) {
                        getSupportFragmentManager().popBackStack();
                    }
                    setupHomeIcon();
                }
                return true;
            }
        });

        searchView.setImeOptions(searchView.getImeOptions() | EditorInfo.IME_FLAG_NO_EXTRACT_UI);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if(item.getItemId() == R.id.mi_favorite) {
            if(!item.isChecked()) {
                item.setIcon(getResources().getDrawable(R.drawable.ic_star));
                viewModel.storeSelectedQuestion();
                item.setChecked(true);
            } else {
                item.setIcon(getResources().getDrawable(R.drawable.ic_star_hollow));
                viewModel.deleteSelectedQuestion();
                item.setChecked(false);
            }
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onQueryTextSubmit(String query) {

        if (query != null) {
            searchView.clearFocus();
            viewModel.submitQuery(query);
        }
        return false;
    }

    @Override
    public boolean onQueryTextChange(String newText) {

        if (newText != null) {
            viewModel.submitQuery(newText);
        }
        return true;
    }

    private void initObservers() {

        viewModel.activeScreen.observe(this, it -> {
            if (it != null && getLifecycle().getCurrentState().isAtLeast(RESUMED)) {
                navigate(it);
            }
        });

        viewModel.connectionStatus().observe(this, it -> {
            if(it != null && !it && getLifecycle().getCurrentState().isAtLeast(RESUMED)) {
                displayConnectionWarning();
            }
        });

        viewModel.isBusy.observe(this, it -> {
            if(it != null) binding.progressIndicator.setVisibility(it? View.VISIBLE : View.GONE);
        });
    }

    private void setupHomeIcon() {
        boolean rootScreen = getSupportFragmentManager().getBackStackEntryCount() == 0;
        toolbar.setNavigationIcon(ResourcesCompat.getDrawable(getResources(),
                                  rootScreen? R.drawable.ic_code : R.drawable.ic_back, null));
    }

    private void displayConnectionWarning() {
        if(binding.toolbarView == null) {
            return;
        }

        Snackbar.make(binding.toolbarView, R.string.error_no_connection, Snackbar.LENGTH_LONG)
                .setAction(R.string.error_no_connection_button_label, it -> {
                    Intent intent = new Intent(Settings.ACTION_SETTINGS);
                    startActivity(intent);
                })
                .show();
    }

    private void initFragments() {
        ActiveScreen activeScreen = viewModel.activeScreen.getValue();
        if(activeScreen == null) {
            return;
        }

        switch (activeScreen) {
            case COLLECTION:
                addMainFragment(new CollectionFragment(), CollectionFragment.TAG);
                break;
            case SEARCH:
                addMainFragment(new SearchResultsFragment(), SearchResultsFragment.TAG);
                break;
            case DETAILS:
                addMainFragment(new DetailsFragment(), DetailsFragment.TAG);
                break;
        }
    }

    private void navigate(ActiveScreen screen) {
        switch (screen) {
            case COLLECTION:
                replaceMainFragment(new CollectionFragment(), CollectionFragment.TAG);
                break;

            case DETAILS:
                invalidateOptionsMenu();
                replaceMainFragment(new DetailsFragment(), DetailsFragment.TAG);
                break;

            case SEARCH:
                replaceMainFragment(new SearchResultsFragment(), SearchResultsFragment.TAG);
                break;
        }
        setupHomeIcon();
    }

    private void addMainFragment(Fragment fragment, String tag) {
        getSupportFragmentManager().beginTransaction()
                .add(R.id.mainFragment, fragment, tag)
                .commit();
    }

    private void replaceMainFragment(Fragment fragment, String tag) {
        if (getSupportFragmentManager().findFragmentByTag(tag) == null) {
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.mainFragment, fragment, tag)
                    .addToBackStack(null)
                    .commit();
        }
    }
}

File: app/src/main/java/io/github/anvell/stackoverview/view/MainActivity.kt
package io.github.anvell.stackoverview.view

import android.arch.lifecycle.Lifecycle.State.RESUMED
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.support.design.widget.Snackbar
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.View
import android.view.inputmethod.EditorInfo
import io.github.anvell.stackoverview.R
import io.github.anvell.stackoverview.enumeration.ActiveScreen
import io.github.anvell.stackoverview.viewmodel.MainViewModel
import kotlinx.android.synthetic.main.activity_main.*
import android.animation.LayoutTransition
import android.widget.LinearLayout


class MainActivity : AppCompatActivity(), SearchView.OnQueryTextListener {

    private lateinit var viewModel: MainViewModel
    private lateinit var searchView: SearchView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        viewModel = ViewModelProviders.of(this).get(MainViewModel::class.java)

        setupHomeIcon(viewModel.activeScreen.value!!)
        initObservers()

        if(savedInstanceState == null) {
            initFragments()
        }

        toolbar.setNavigationOnClickListener {
            viewModel.clearSelectedQuestion()
            navigate(ActiveScreen.SEARCH)
        }

        if (Build.VERSION.SDK_INT >= 21) {
            toolbar.elevation = 4f
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        viewModel.clearSelectedQuestion()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)

        menu?.let {
            val mi = menu.findItem(R.id.mi_search_view)
            searchView = mi.actionView as SearchView
            val searchBar = searchView.findViewById(R.id.search_bar) as LinearLayout
            searchBar.layoutTransition = LayoutTransition()

            searchView.setOnQueryTextListener(this)
            searchView.setOnCloseListener {
                viewModel.selectedQuestion.value?.let {
                    if(viewModel.activeScreen.value != ActiveScreen.DETAILS)
                        navigate(ActiveScreen.DETAILS)
                }
                false
            }

            searchView.imeOptions = searchView.imeOptions or EditorInfo.IME_FLAG_NO_EXTRACT_UI
        }
        return true
    }

    override fun onQueryTextSubmit(query: String?): Boolean {
        query?.let {
            searchView.clearFocus()
            searchView.onActionViewCollapsed()
            viewModel.submitQuery(query)
        }
        return false
    }

    override fun onQueryTextChange(newText: String?): Boolean {
        newText?.let {
            viewModel.submitQuery(newText)
        }
        return true
    }

    private fun initObservers() {
        viewModel.activeScreen.observe(this, Observer {
            it?.let {
                if(lifecycle.currentState.isAtLeast(RESUMED)) {
                    navigate(it)
                }
            }
        })

        viewModel.connectionStatus().observe(this, Observer {
            it?.let {
                if(!it && lifecycle.currentState.isAtLeast(RESUMED)) {
                    displayConnectionWarning()
                }
            }
        })

        viewModel.isBusy.observe(this, Observer {
            progressIndicator.visibility = when (it!!) {
                true -> View.VISIBLE
                false -> View.GONE
            }
        })
    }

    private fun setupHomeIcon(screen: ActiveScreen) {
        toolbar.navigationIcon = when (screen) {
            ActiveScreen.DETAILS -> ResourcesCompat.getDrawable(resources, R.drawable.ic_back, null)
            ActiveScreen.SEARCH -> ResourcesCompat.getDrawable(resources, R.drawable.ic_code, null)
        }
    }

    private fun displayConnectionWarning() {
        Snackbar.make(toolbar, R.string.error_no_connection, Snackbar.LENGTH_LONG).apply {
            setAction(R.string.error_no_connection_button_label, {
                val intent = Intent(Settings.ACTION_SETTINGS)
                startActivity(intent)
            })
            show()
        }
    }

    private fun initFragments() {
        val (fragment, tag) = when (viewModel.activeScreen.value) {
            ActiveScreen.DETAILS -> Pair(DetailsFragment(), DetailsFragment.TAG)
            ActiveScreen.SEARCH -> Pair(SearchResultsFragment(), SearchResultsFragment.TAG)
            else -> return
        }

        supportFragmentManager.beginTransaction()
                .add(R.id.mainFragment, fragment, tag)
                .commit()
    }

    private fun navigate(screen: ActiveScreen): Boolean {
        when(screen) {
            ActiveScreen.DETAILS -> {
                searchView.onActionViewCollapsed()
                supportFragmentManager.beginTransaction()
                        .replace(R.id.mainFragment, DetailsFragment(), DetailsFragment.TAG)
                        .addToBackStack(null)
                        .commit()
            }

            ActiveScreen.SEARCH -> {
                with(supportFragmentManager) {
                    if (backStackEntryCount > 0) {
                        popBackStack()
                    }
                }
            }
        }
        setupHomeIcon(screen)
        return true
    }
}
